### DISCUSSION TITLE
THE most valuable thing for the TW project?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5392] [General]
A very open question:

What one thing or activity would contribute most to the TW *project*?

<hr>

My own answer is; A practical solution for the community to be able to share plugins etc. By "*practical* solution" I mean something that the average user can actually and easily use.

(I'm "lucky" to presently have less time than ever to work on tiddly stuff because if I did have the time I would suffer tremendously from the tiddlyspot shutdown. It was truly my playground and a necessity for my contributions.)

**Comment by CodaCodr** [comment]
THE most valuable thing for the TW project? 
* The curator - Jeremy.

What one thing or activity would contribute most to the TW project? 
* A better forum - competing with GG can't be difficult. THIS is better than GG.

(Those are very different questions, Mat.)

**Comment by saqimtiaz** [comment]
Also sharing plugins makes me think of a community plugin library rather than a hosting service like TiddlySpot.

Github saver is pretty good now with the recent bug fixes and I think @kookma has a good guide for setting it up. And you just need to set it up once per computer for all your wikis. Might not be ideal but I think that is best option at present.

**Comment by kookma** [comment]
Thank you @saqimtiaz - Well there are several tutorials to setup Tiddlywiki on GitHub. See for example https://github.com/kookma/TW5-GitHub-Saver
@twMat  - I think as you have GitHub account it is not difficult to setup your wikis on GitHub pages!

**Reply by saqimtiaz** [reply]
@kookma maybe update this part:
`Target branch=use default (master)` at https://kookma.github.io/TW5-GitHub-Saver/#Quick%20Setup

**Reply by kookma** [reply]
Sure, I will update - Thank you

**Comment by morosanuae** [comment]
I know what @twMat wants because I want it too. There are a lot of useful plugins out there but they are not easy to find. We need a CENTRAL public place/method for sharing plugins. An website or a plugin library or something. Where all the plugins can adhere to some quality/technical standards. Maybe also to provide ability for users to evaluate them, vote them, categorize them, share useful experience or provide technical support. Something organized not as it is now, more like a jungle. Anyway, you get the point...

**Reply by kookma** [reply]
> I know what @twMat wants because I want it too. There are a lot of useful plugins out there but they are not easy to find. We need a CENTRAL public place/method for sharing plugins. An website or a plugin library or something. Where all the plugins can adhere to some quality/technical standards. Maybe also to provide ability for users to evaluate them, vote them, categorize them, share useful experience or provide technical support. Something organized not as it is now, more like a jungle. Anyway, you get the point...

@NicolasPetton and @saqimtiaz  - I highly support having such community plugin library!

**Reply by NicolasPetton** [reply]
I've been recently playing with building a community-maintained plugin library. It's not ready for prime time yet, but I hope to be able to share something soon.

**Reply by kookma** [reply]
Hi @NicolasPetton - I am ready to join and add my working plugins to the mentioned library!

**Comment by joshuafontany** [comment]
Following this with interest. :)

**Comment by Jermolene** [comment]
Thanks @twMat your question is phrased in a useful way.

As a community, we have an opportunity to work together to achieve our collective aims, and thus a vital question is how we prioritise the different things that we could choose to do according to their value to the community and our goals. We need to be able to talk about priorities without getting too distracted by implementation details.

In terms of community infrastructure, there are several substantial things we lack, and a few that could be improved:

* Community plugin library (actually not just for plugins, also palettes, themes, tips and snippets)
* Community hosting service (ie a free TW5 replacement for TiddlySpot)
* An easier means for community members to update the content on tiddlywiki.com
* Better discussion forum than Google Groups

The community plugin library and hosting service might sound like they could be the same thing, but they have very different constraints: the plugin library is for shared community resources for which we take some collective responsibility, while the hosting service would need to be available for anyone to host any type of content, with no collective responsibility.

I have also come to believe that creating a community plugin library would have the most impact on the project, but I'm interested to hear other views. If so, then we should start a new discussion thread to get into more detail. At this point, that means pulling together a list of the goals and requirements for the effort.

The process I'm describing might sound a bit lumbering but it's important to recognise that this is only tangentially a technology implementation problem, I think it's primarily about understanding the needs of the different voices in the community.

**Reply by morosanuae** [reply]
@Jermolene Sounds right! Why not creating a new discussion on this matter then? We should get started right away, at least with "**TiddlyWiki Community Resources Library**". I think a lot of people will be very happy and would like to share their thoughts.

**Comment by twMat** [comment]
>What one thing or activity would contribute most to the TW project?

To my surprise, the discussion thus far indicates more or less a consensus on the matter.

@kookma (and @saqimtiaz ) , I've posted some [issues](https://github.com/kookma/TW5-GitHub-Saver/issues/3) for your setup. So far it is unfortunately far from anything that an average user can easily set up. (Side note: A *video* is probably the clearest way to show how to do it.)

@NicolasPetton , a community maintained plugin library sounds great but, as Jeremy indicates, where would the plugins be stored? 

I agree with @Jermolene 's suggestion for a new thread and:

>pulling together a list of the goals and requirements for the effort.

@Jermolene - do you think you could start this thread? That gives it more weight than if e.g I do it and the objective would not be "misphrased".

Thanks everyone.

**Reply by NicolasPetton** [reply]
> @NicolasPetton , a community maintained plugin library sounds great but, as Jeremy indicates, where would the plugins be stored?

If it's an unofficial plugin library, I could provide the infrastructure.
If' it's an official plugin library, then it's another matter.

The plugin library I'm working on is based on the following principles:

* Curated – Plugin submission would have to be validated
* Up-to-date – Plugins built from upstream source only
* Extensible – Contribute new recipes, and we'll build the plugins.
* Automatic updates – Version bumps in plugin.info files result in new plugins
* Stable – It only contains stable releases of plugins

**Reply by Jermolene** [reply]
Hi @NicolasPetton thank you for the offer. I've tried to explain that I don't think the provision of the infrastructure is the problem here. As we discovered in our earlier private conversation, we have slightly differing views of the scope and goals of the plugin library. That's not in itself a problem of course, but it does highlight that if this is to be a community project then we need to start by discussing the goals, scope etc. rather than jumping straight to the implementation.

**Reply by Jermolene** [reply]
> @Jermolene - do you think you could start this thread? That gives it more weight than if e.g I do it and the objective would not be "misphrased".

I've started the discussion over at #5399.

**Reply by NicolasPetton** [reply]
> I don't think the provision of the infrastructure is the problem here. 

I don't think so either, I was just answering @twMat.

>  we have slightly differing views of the scope and goals of the plugin library. 

I'm totally fine with that. I think (but I'm not sure) that you're thinking about an "official" community plugin library. That's not necessarily my goal (though I'm not at all against that, of course). 

My initial goal was simply to answer a problem I currently have myself: how do I distribute & keep up-to-date my plugins? The scope grew when I realized that others might have the same issue and could want to publish their packages the same way.

> That's not in itself a problem of course, but it does highlight that if this is to be a community project then we need to start by discussing the goals, scope etc. rather than jumping straight to the implementation.

Sure, I'll open a new thread.

However, I do not necessarily agree with the idea of discussing and waiting for a plan to be laid out to start working on the implementation. Sure, it is sometimes necessary, but building an MVP based on a couple of ideas and letting people use it to see if it sticks can also be a good approach IMHO.

In this specific case, my idea was to see if a plugin library that would borrow the concepts of [MELPA](https://melpa.org) would work for the TiddlyWiki community.

**Comment by silvyn** [comment]
Dreaming of a replacement for Google Groups in 2021: is it a dream?
We know it is, it's a sea snake with multiple discussions on GG... and yet it's getting worse and worse and I find it relevant !

For five years now, I have maintained Discourse for the French community. So frustrated not to have the same tools for everyone and say good by GG !

**Comment by CodaCodr** [comment]
@silvyn  I mentioned Discourse before but it went (largely) nowhere. In another "life", I'm a mod/leader/ambassador using Discourse. It _does_ have its own issues... but really, nothing like as bad as GG.

**Reply by jussiarpalahti** [reply]
From my new user perspective I'd like to 👍 Discourse. I had a hard time with GG. Searching in it and in the alternative archive was kludgy at best. Also, trying to find new posts is conservations is frustrating.

**Comment by jussiarpalahti** [comment]
After many years I've returned to using Tiddlywiki. I have only just started. I don't have clear idea yet of Tiddlywiki architecture nor how to make plugins currently. So bear this in mind for what I am musing below.

I have lately gotten back to JavaScript development thanks solely to modern JavaScript and ES modules coming to browsers. With HTTP 2 and caching it's actually *both* easy and fast to load JS straight from NPM, or from services like Skypack and Unpkg. Thus I'd like to simply write plugins as JS modules and import them straight to Tiddly. I could even develop them by coding straight into a tiddler inside browser.

Technically this is doable by something as simple as:

```javascript
import {registry} from "TiddlyWiki/plugins"
registry.register(my_plugin_object)
```

This can be either a JS file published by NPM or evaluated from tiddler contents (highly dangerous) with eval inside browser.

Since I'm currently ignorant enough not to know the feasibility of my idea I wanted to just note for the community. This method is how **I** like to write JS these days. I do realise it's not universal. Backwards compatibility and need to support older browsers would make this harder to implement, though not impossible.

**Reply by Jermolene** [reply]
> I have lately gotten back to JavaScript development thanks solely to modern JavaScript and ES modules coming to browsers. With > Since I'm currently ignorant enough not to know the feasibility of my idea I wanted to just note for the community. This method is how **I** like to write JS these days. I do realise it's not universal. Backwards compatibility and need to support older browsers would make this harder to implement, though not impossible.

Indeed, those are the kind of modern JS features I'd like to be able to use. I do a little non-TW5 JS programming where I'm able to use the new features and love a lot of them. I'm interested in [Deno](https://deno.land/) too. As it happens I'm not at all a fan of TypeScript.

**Reply by jussiarpalahti** [reply]
Glad to hear, @Jermolene!

By the by, [TypeScript supports JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) for some parts of its functionality. When comments have certain (possibly even relatively standard) structure TypeScript can understand regular JavaScript syntax and perform type checking. This works in VS Code out of the box these days, I believe.

**Reply by CodaCodr** [reply]
> As it happens I'm not at all a fan of TypeScript.

Dreadful. Coding == Composing (CodaCoder - geddit?) I don't want that kinda guidance or anyone restricting the movements of my fingers over a fretboard. And type-safety? you don't need it ([link](https://medium.com/javascript-scene/the-typescript-tax-132ff4cb175b))

I've been coding against selenium/webdriver for four years at this point using Groovy/Java - types are driving me nuts on a daily basis. I break out into JavaScript "all the time"...

JavaScript -> Freedom.
Java (types) -> Thought police.

/rant (sorry)

**Comment by TiddlyTweeter** [comment]
@twMat wrote ...
> A very open question:
> 
> What one thing or activity would contribute most to the TW _project_?

I have read the discussion with interest, One thing that is perhaps not quite yet foregrounded enough and very relevant to your query above is ... a need for a resource that better matches user END NEEDS to simple click-and-get ONE-STOP solutions???

In my mind is users _not_ interested in coding at all. Rather only on functional (end) needs that are domain/interest focused. So I'm thinking: _How do I find_ the best (finished) TW (edition?) for an Anthropologist to collect field notes? Or an Artist to showcase their painting?

---

The point I'm trying to get at, I hope relevant to the discussion, is that stepping towards a more centralized system are we aiming to ease access to "Components" to build solutions OR ease access to full (ready-to-run) "Solutions"?

**Reply by twMat** [reply]
@TiddlyTweeter , I agree with what you say about ready made editions *but* a functional infrastructure to make them available is still a prerequisite for that. I'd even think a user friendly "distribution systems" is exactly the kind of thing that might *encourage* users to create such editions.

**Reply by TiddlyTweeter** [reply]
> @TiddlyTweeter , I agree with what you say about ready made editions _but_ a functional infrastructure to make them available is still a prerequisite for that. I'd even think a user friendly "distribution systems" is exactly the kind of thing that might _encourage_ users to create such editions.

@twMat. I think we are in agreement in the end :-). I do agree that evolving a better infrastructure would help a lot. I think my _residual concern_ is only that the "ultimate" aim is to get our great tool used more. And I think _replete_, "off-the-shelf" applications matter to that. So I'm interested in innovation that somewhere has that clearly in mind.

**Comment by twMat** [comment]
Side note: I posted [Please help to refine this (working!) Google Drive backend #5482](https://github.com/Jermolene/TiddlyWiki5/discussions/5482) which I believe should enable users (who have a google account) to publish plugins etc via the following idea:

There's a central wiki hosted somewhere that has the (requested) "GoogleDriveTWbackend" plugin installed. The wiki can also have its own automated listing of plugins thanks to Googles excellent Forms which basically lets you create "questionnaires" that anyone can fill in and *all* replies are automatically collected into a Google Sheet. (BTW, the form can be presented in that very public wiki). So... this sheet data is fetched by that central wiki and presented. The visitor can select which plugin he wants and it is fetched from the plugin creators sheet with tiddlers. 

Complex yes, complicated not very.

**Reply by CodaCodr** [reply]
Me like. That would be one helluva showcase, if nothing else...

**Reply by twMat** [reply]
Didn't think of that aspect but, yes, you're absolutely right. We could showcase *really* cool stuff because there is a very *wide* range of integration between various Google services and, specifically, Google Sheets. And the "JSON-fetching" parts in Siniy-Kits solution that need extraction are even really small, so a small plugin would really open up a "world" to the user.

I have a momentary breather in my work so if anyone who knows some JS is interested in collaborating on this, I'm all up for it.

**Comment by linonetwo** [comment]
I think it will be "API documentation with example and typing".

Learning usage of API is blocking me to create more complex plugin!
